In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)
#to use persistent storage

Mounted at /content/drive


In [ ]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Thu_Jun__6_02:18:23_PDT_2024
Cuda compilation tools, release 12.5, V12.5.82
Build cuda_12.5.r12.5/compiler.34385749_0


In [ ]:
import sys
import os
import json

# Append the directory containing your package
#sys.path.append(os.path.abspath('/content/drive/MyDrive/mRAG_and_MSRS_source'))
#data_path = "/content/drive/MyDrive/mRAG_and_MSRS_source"

#location of corpus
jsonl_dir = '/content/drive/MyDrive/mRAG_and_MSRS_source/jsonl_collections'

index_dir = '/content/drive/MyDrive/mRAG_and_MSRS_source/index'
print(os.path.isdir(jsonl_dir))

#corpus_chunk_dir = os.path.join(data_path, "jsonl_collections") # storing the converted corpus entries
#index_dir = os.path.join(data_path, "index") #  storing the index components from pyserini


True


In [ ]:
#NOTE: harmless runtime restart will take place, just need to reexecute this block

!pip -q install condacolab
import condacolab; condacolab.install()

#creating two envs because these were technically 2 apps with 2 different requirements.txt(dependencies) in the mRAG repository
!mamba create -y -n retriever python=3.11
!mamba create -y -n train_and_test python=3.11


✨🍰✨ Everything looks OK!

Looking for: ['python=3.11']

conda-forge/linux-64                                        Using cache
conda-forge/noarch                                          Using cache
Transaction

  Prefix: /usr/local/envs/retriever

  Updating specs:

   - python=3.11


  Package                Version  Build                 Channel           Size
────────────────────────────────────────────────────────────────────────────────
  Install:
────────────────────────────────────────────────────────────────────────────────

  + tzdata                 2025b  h78e105d_0            conda-forge     Cached
  + ca-certificates   2025.11.12  hbd8a1cb_0            conda-forge     Cached
  + _libgcc_mutex            0.1  conda_forge           conda-forge     Cached
  + libgomp               15.2.0  he0feb66_14           conda-forge     Cached
  + _openmp_mutex            4.5  2_gnu                 conda-forge     Cached
  + libgcc                15.2.0  he0feb66_14           conda-fo

In [ ]:
#retriever server dependency
!mamba run -n retriever pip install uvicorn



#!mamba run -n retriever pip install conda-pack
#!mamba run -n train_and_test pip install conda-pack

#ensure installation of certain files related to cuda 12.5 the one that comes with this runtime despite repo 12.6 suggestion
#!pip install --index-url https://download.pytorch.org/whl/cu124 torch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1

#installing the requirements for the mRAG project github repo

!mamba run -n retriever pip install -r  /content/drive/MyDrive/mRAG_and_MSRS_source/retriever_setup/requirements.txt
#!PIP_CACHE_DIR=/content/drive/MyDrive/mRAG_and_MSRS_source/pip_cache_retriever mamba run -n retriever pip install -r  /content/drive/MyDrive/mRAG_and_MSRS_source/retriever_setup/requirements.txt
!mamba run -n retriever pip install fastapi





  Using cached https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.8.0/en_core_web_sm-3.8.0-py3-none-any.whl (12.8 MB)
  Using cached accelerate-1.2.0-py3-none-any.whl.metadata (19 kB)
  Using cached aiohappyeyeballs-2.4.4-py3-none-any.whl.metadata (6.1 kB)
  Using cached aiohttp-3.11.11-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (7.7 kB)
  Using cached aiosignal-1.3.2-py2.py3-none-any.whl.metadata (3.8 kB)
  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)
  Using cached asttokens-3.0.0-py3-none-any.whl.metadata (4.7 kB)
  Using cached async_timeout-5.0.1-py3-none-any.whl.metadata (5.1 kB)
  Using cached attrs-24.3.0-py3-none-any.whl.metadata (11 kB)
  Using cached beir-2.0.0.tar.gz (53 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing 

In [ ]:
!mamba run -n train_and_test pip install -r  /content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training/requirements.txt


  Using cached aiosignal-1.3.2-py2.py3-none-any.whl.metadata (3.8 kB)
  Using cached aiosqlite-0.21.0-py3-none-any.whl.metadata (4.3 kB)
  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)
  Using cached banks-2.1.2-py3-none-any.whl.metadata (12 kB)
  Using cached click-8.1.8-py3-none-any.whl.metadata (2.3 kB)
  Using cached colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)
  Using cached dataclasses_json-0.6.7-py3-none-any.whl.metadata (25 kB)
  Using cached Deprecated-1.2.18-py2.py3-none-any.whl.metadata (5.7 kB)
  Using cached dill-0.3.8-py3-none-any.whl.metadata (10 kB)
  Using cached dirtyjson-1.0.8-py3-none-any.whl.metadata (11 kB)
  Using cached distro-1.9.0-py3-none-any.whl.metadata (6.8 kB)
  Using cached filetype-1.2.0-py2.py3-none-any.whl.metadata (6.5 kB)
  Using cached frozenlist-1.5.0-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)
  Using cached httpx-0.28.1-py3-none-any.whl.metadata

In [ ]:
#installed at a separate point removed from requirements.txt file, because torch has to be guaranteed to be built before
!mamba run -n train_and_test pip install --no-build-isolation flash-attn==2.7.4.post1

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 94.2 MB/s  0:00:00
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for flash-attn: filename=flash_attn-2.7.4.post1-cp311-cp311-linux_x86_64.whl size=187815463 sha256=d944fc7d2f962bce83fc4708c2fc0c21eaf8255962a0b350ae919362a51b7ef2
  Stored in directory: /root/.cache/pip/wheels/3d/88/d8/284b89f56af7d5bf366b10d6b8e251ac8a7c7bf3f04203fb4f
Successfully built flash-attn



In [ ]:
#CACHE the entire mamba models
#!mamba run -n retriever conda-pack -o /content/drive/MyDrive/mRAG_and_MSRS_source/conda_envs/retriever.tar.gz
#!mamba run -n train_and_test conda-pack -o /content/drive/MyDrive/mRAG_and_MSRS_source/conda_envs/train_and_test.tar.gz

In [ ]:
#!mamba run -n train_and_test pip install conda-pack
#!mamba run -n train_and_test conda-pack -o /content/drive/MyDrive/mRAG_and_MSRS_source/conda_envs/train_and_test.tar.gz

In [ ]:
#Retrieve cached mamba models

# ############ MSRS PORTION (https://huggingface.co/datasets/yale-nlp/MSRS)

#Indexing Task 1 (Always Run)

In [ ]:
######## In place of using datamorgana using MSRS Benchmark
######## due to proprietary nature of datamorgana.
######## Gold answers in MSRS paper are groundtruth answers
######## morphing MSRS set/benchmark to be similar in format to datamorgana set/benchmark


!pip install datasets
import pickle

from datasets import load_dataset

story_corpus = load_dataset("yale-nlp/MSRS", "story-corpus", split="corpus")

#for use in the retriever_setup code in mRAG app, needs a doc_ids pickle
doc_ids = [row["id"] for row in story_corpus]
with open(os.path.join("/content/drive/MyDrive/mRAG_and_MSRS_source/index", "doc_ids.pkl"), "wb") as f:
    pickle.dump(doc_ids, f)


meeting_corpus = load_dataset("yale-nlp/MSRS", "meeting-corpus", split="corpus")
'''
this is a list of corpus'
{
    "id": // Unique ID for the document
    "text": // Document text
}
'''


print(type(story_corpus))
#'contents' is expected instead of 'text' by pyserini basic indexer so rename that field

story_corpus = story_corpus.rename_columns({"text": "contents"})

#Any necessary preprocessing?

#exporting the corpus to jsonl
story_corpus.to_json(jsonl_dir+"/story_corpus.jsonl", lines=True)

print(os.listdir(jsonl_dir))
#meeting_corpus.to_json("./jsonl_collections/meeting_corpus.jsonl", lines=True)

#story_corpus.to_json(data_path + "/story_corpus.jsonl", lines=True)
#meeting_corpus.to_json(data_path + "/meeting_corpus.jsonl", lines=True)


  Using cached httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)
  Using cached anyio-4.12.0-py3-none-any.whl.metadata (4.3 kB)
  Using cached httpcore-1.0.9-py3-none-any.whl.metadata (21 kB)
  Using cached h11-0.16.0-py3-none-any.whl.metadata (8.3 kB)
  Using cached shellingham-1.5.4-py2.py3-none-any.whl.metadata (3.5 kB)
  Using cached typing_extensions-4.15.0-py3-none-any.whl.metadata (3.3 kB)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
  Using cached aiohappyeyeballs-2.6.1-py3-none-any.whl.metadata (5.9 kB)
  Using cached six-1.17.0-py2.py3-none-any.whl.metadata (1.7 kB)
  Using cached click-8.3.1-py3-none-any.whl.metadata (2.6 kB)
Using cached httpx-0.28.1-py3-none-any.whl (73 kB)
Using cached httpcore-1.0.9-py3-none-any.whl (78 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.9/16.9 MB 215.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.7/47.7 MB 74.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.6/806.6 k

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md: 0.00B [00:00, ?B/s]

corpus.jsonl: 0.00B [00:00, ?B/s]

Generating corpus split:   0%|          | 0/1138 [00:00<?, ? examples/s]

corpus.jsonl: 0.00B [00:00, ?B/s]

Generating corpus split:   0%|          | 0/231 [00:00<?, ? examples/s]

<class 'datasets.arrow_dataset.Dataset'>


Creating json from Arrow format:   0%|          | 0/2 [00:00<?, ?ba/s]

['story_corpus.jsonl']


In [ ]:
# print(len(story_qa))
# print(len(list(story_qa)))
# print(story_qa)

# print(story_qa['train'])
# print(story_qa['test'])

print(len(story_corpus))
print(story_corpus)

1138
Dataset({
    features: ['id', 'contents'],
    num_rows: 1138
})


# Chunking Corpus (to tokens rather than what was done in Assignment 1)
# NOTE: retriever_setup app in the repository is to setup retriever, but is mostly to facilitate the SparseRetriever model, BM25Retriever just functions off of an index

## Changes to files:
#retriever_setup/fine_web_chunking.py
#retreiver_setup/constants.py
#retrieval/retrievers.py ( for BM25 changed referenced index and the format of #output)


Because BM25 is hardcoded to reference an index, need to generate that using MSRS Corpus

#Indexing Task 2

In [ ]:
# ### Pyserini require updating java version ###
# !apt-get install openjdk-21-jdk-headless -qq > /dev/null
# os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-21-openjdk-amd64"
# !update-alternatives --set java /usr/lib/jvm/java-21-openjdk-amd64/jre/bin/java
# !java -version


# #!pip install pyserini

E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/universe/o/openjdk-21/openjdk-21-jre-headless_21.0.8%2b9%7eus1-0ubuntu1%7e22.04.1_amd64.deb  404  Not Found [IP: 91.189.92.23 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/universe/o/openjdk-21/openjdk-21-jdk-headless_21.0.8%2b9%7eus1-0ubuntu1%7e22.04.1_amd64.deb  404  Not Found [IP: 91.189.92.23 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
update-alternatives: error: alternative /usr/lib/jvm/java-21-openjdk-amd64/jre/bin/java for java not registered; not setting
openjdk version "17.0.16" 2025-07-15
OpenJDK Runtime Environment (build 17.0.16+8-Ubuntu-0ubuntu122.04.1)
OpenJDK 64-Bit Server VM (build 17.0.16+8-Ubuntu-0ubuntu122.04.1, mixed mode, sharing)


In [ ]:
import os

# 1. Refresh the package list (Crucial step to fix 404 errors)
print("Updating package lists...")
!apt-get update > /dev/null

# 2. Install Java 21 (with --fix-missing to handle connection hiccups)
print("Installing Java 21...")
!apt-get install -y openjdk-21-jdk-headless --fix-missing > /dev/null

# 3. Verify the installation
print("Verifying installation...")
!java -version

# 4. Set the Environment Variables for THIS Python session
# (This tells Pyserini exactly where the new Java is)
java_home = "/usr/lib/jvm/java-21-openjdk-amd64"
os.environ["JAVA_HOME"] = java_home
os.environ["JVM_PATH"] = f"{java_home}/lib/server/libjvm.so"

print(f"\nSUCCESS: JAVA_HOME set to {os.environ['JAVA_HOME']}")

Updating package lists...
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Installing Java 21...
Verifying installation...
openjdk version "21.0.9" 2025-10-21
OpenJDK Runtime Environment (build 21.0.9+10-Ubuntu-122.04)
OpenJDK 64-Bit Server VM (build 21.0.9+10-Ubuntu-122.04, mixed mode, sharing)

SUCCESS: JAVA_HOME set to /usr/lib/jvm/java-21-openjdk-amd64


In [ ]:
print(os.listdir(index_dir))

['write.lock', '_2.nvd', '_2.nvm', '_2_Lucene90_0.dvd', '_2_Lucene90_0.dvm', '_2.fdx', '_2.fdm', '_2.fdt', '_2.tvx', '_2.tvm', '_2.tvd', '_2_Lucene99_0.tmd', '_2_Lucene99_0.tim', '_2_Lucene99_0.tip', '_2_Lucene99_0.doc', '_2_Lucene99_0.pos', '_2.fnm', '_2.si', 'segments_3', 'doc_ids.pkl']


#Indexing Task 3

In [ ]:
if os.path.isdir(index_dir) and os.listdir(index_dir):
    print('index folder is not empty, found processed files.')
else:
    # Call the Pyserini indexer function here
    #for jsonl_file in os.listdir(jsonl_dir):
    !python -m pyserini.index.lucene --collection JsonCollection --input {jsonl_dir} --index {index_dir} --generator DefaultLuceneDocumentGenerator --threads 8 --storePositions --storeDocvectors --storeRaw


index folder is not empty, found processed files.


NOTE: number of documents significantly small in MSRS-Story (1138), will step this up using hotpotqa dataset in another test

#Creating Chunks(Pre-processing) Conforms to size and overlap in fine_web_chunking.py

In [ ]:
#Pre-processed chunking, file required in sparse_retrieval_cpu.py script

#Question: what is the relevance of chunk size and overlap to size of text?

#NOTE: chunk sizes match those specified in fine_web_chunking.py SentenceSplitter
def make_chunks(text, max_words=512, overlap=80):
    if not text:
        return
    words = text.split()
    step = max_words - overlap
    if step <= 0:
        step = max_words
    for i in range(0, len(words), step):
        yield " ".join(words[i:i+max_words])

out_path = "/content/drive/MyDrive/mRAG_and_MSRS_source/chunked_data/story_corpus_chunked.jsonl"
with open(out_path, "w", encoding="utf-8") as f:
    for row in story_corpus:
        doc_id = row["id"]
        text = row["contents"]
        for j, chunk in enumerate(make_chunks(text)):
            rec = {
                "chunk_id": f"{doc_id}#{j}",
                "contents": chunk
            }
            f.write(json.dumps(rec, ensure_ascii=False) + "\n")

In [ ]:
import pickle
#for use in the retriever_setup code in mRAG app, needs a doc_ids pickle
doc_ids = [row["id"] for row in story_corpus]
with open(os.path.join("/content/drive/MyDrive/mRAG_and_MSRS_source/index", "doc_ids.pkl"), "wb") as f:
    pickle.dump(doc_ids, f)

# ############# mRAG PORTION (https://github.com/muktac5/CIIR-LiveRAG/tree/main)

###  SERVER INITIALIZATION, hosted on 127.0.0.1, localhost ################

In [ ]:
######### Start the Retriever server  ( pure CPU )
######### Initially Retrievesd from FineWeb Dataset, Modified to retrieve from MSRS Corpus

#huggingface access token

added to files:

./retriever_setup/eval_sparse.py
./retriever_setup/scaling_retriever/modeling/llm_encoder.py


#from huggingface_hub import login


Reason for providing token : Is this because for sparse retriever requires llama2 and llama2 requires user license agreement

In [ ]:
######################################################

When

In [ ]:
# #UPDATE: made comment in sparse_retrieval_cpu.py to enforce 'cpu' use to conserve cuda for other models:
# #device = "cpu"

# !echo "127.0.0.1" > /content/drive/MyDrive/mRAG_and_MSRS_source/retriever_log/retriever.log
# !echo "8000" >> /content/drive/MyDrive/mRAG_and_MSRS_source/retriever_log/retriever.log


# #nohup is necessary to run in the background and have other cells as well(nonblocking)

# !PYTHONPATH=/content/drive/MyDrive/mRAG_and_MSRS_source/retriever_setup \
#   nohup mamba run -n retriever uvicorn \
#   --app-dir /content/drive/MyDrive/mRAG_and_MSRS_source/retriever_setup \
#   sparse_retrieval_cpu:app --host 127.0.0.1 --port 8000 \
#   >> /content/drive/MyDrive/mRAG_and_MSRS_source/retriever_log/retriever.log 2>&1 & echo $!



5920


In [ ]:
#Reclaim retriever server memory
# !pkill uvicorn

In [ ]:
# !pkill -9 uvicorn


In [ ]:
######### Kill the Retriever server

In [ ]:
# #killing the retriever
# !kill -TERM 35084
# #close the associated port
# !fuser -k 8000/tcp

/bin/bash: line 1: kill: (35084) - No such process


In [ ]:
# #NOTE: the cell output seems to be a lowball, but might be true, these are in kilobytes

# #NOTE: in reality seems like it takes more
# #5.2GB is the estimate from jump

# #resources stats for launched process(vram)
# !ps -p 30584 -o pid,ppid,cmd,%mem,rss,vsz
# #Uses 9.5GB on the GPU, but was supposed to use CPU right?
# #Uses a little bit of ram

    PID    PPID CMD                         %MEM   RSS    VSZ


In [ ]:
######### Start the Agent server

In [ ]:
#NOTE: attempting to limit system ram with --swap-space 0 or --swap-space 1 causes crash
#0.5 works for gpu memory utilization, but

!echo "127.0.0.1" > /content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log/agent.log
!echo "8100" >> /content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log/agent.log




#NOTE: shrinking context and kv cache datatype 16->8 allows for fitting in a smaller portion of the gpu, aka, .15 of the gpu
# takes time for the model to get loaded into the gpu

!PYTHONPATH=/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training \
  VLLM_LOGGING_LEVEL=INFO\
  nohup mamba run -n train_and_test vllm serve "Qwen/Qwen2.5-7B-Instruct-GPTQ-Int4" --host 127.0.0.1 --port 8100 --uvicorn-log-level info  --download_dir /content/drive/MyDrive/mRAG_and_MSRS_source/vllm_cache --gpu-memory-utilization 0.75 --swap-space 0 --enforce-eager  \
  >> /content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log/agent_runtime.log 2>&1 & echo $!

12367


In [ ]:
!pkill vllm

In [ ]:
# %%bash
# set -euo pipefail
# mkdir -p /content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log

# # discovery file (2 lines host/port)
# printf "127.0.0.1\n8100\n" > /content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log/agent.log

# # runtime log
# RLOG=/content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log/agent_runtime.log
# : > "$RLOG"

# export PYTHONPATH=/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training
# export VLLM_LOGGING_LEVEL=INFO

# nohup mamba run -n train_and_test vllm serve "Qwen/Qwen2.5-0.5B-Instruct-GPTQ-Int4" \
#   --host 127.0.0.1 --port 8100 \
#   --uvicorn-log-level info \
#   --download_dir /content/drive/MyDrive/mRAG_and_MSRS_source/vllm_cache \
#   --gpu-memory-utilization 0.15 \
#   --max-model-len 10240 --swap-space 0 --enforce-eager \
#   >> "$RLOG" 2>&1 & echo $!

6014


In [ ]:
!curl -s http://127.0.0.1:8100/v1/models | jq

curl: /usr/local/lib/libcurl.so.4: no version information available (required by curl)
{
  "object": "list",
  "data": [
    {
      "id": "Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4",
      "object": "model",
      "created": 1764707027,
      "owned_by": "vllm",
      "root": "Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4",
      "parent": null,
      "max_model_len": 10240,
      "permission": [
        {
          "id": "modelperm-a35e0f6af89d41b482b6ccbdf2f5afbb",
          "object": "model_permission",
          "created": 1764707027,
          "allow_create_engine": false,
          "allow_sampling": true,
          "allow_logprobs": true,
          "allow_search_indices": false,
          "allow_view": true,
          "allow_fine_tuning": false,
          "organization": "*",
          "group": null,
          "is_blocking": false
        }
      ]
    }
  ]
}


In [ ]:
!ps -ef | grep vllm

root       12367       1  0 20:20 ?        00:00:00 /usr/local/bin/python /usr/local/bin/mamba run -n train_and_test vllm serve Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4 --host 127.0.0.1 --port 8100 --uvicorn-log-level info --download_dir /content/drive/MyDrive/mRAG_and_MSRS_source/vllm_cache --gpu-memory-utilization 0.75 --max-model-len 10240 --swap-space 0 --enforce-eager
root       12387   12372  3 20:20 ?        00:00:09 /usr/local/envs/train_and_test/bin/python3.11 /usr/local/envs/train_and_test/bin/vllm serve Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4 --host 127.0.0.1 --port 8100 --uvicorn-log-level info --download_dir /content/drive/MyDrive/mRAG_and_MSRS_source/vllm_cache --gpu-memory-utilization 0.75 --max-model-len 10240 --swap-space 0 --enforce-eager
root       13510    1998  0 20:24 ?        00:00:00 /bin/bash -c ps -ef | grep vllm
root       13512   13510  0 20:24 ?        00:00:00 grep vllm


#Example access to server using api

In [ ]:
import requests
resp = requests.post(
    "http://127.0.0.1:8100/v1/chat/completions",
    json={
        "model": "Qwen/Qwen2.5-7B-Instruct-GPTQ-Int4",
        "messages": [{"role":"user","content":"Write a paragraph about books"}],
        "temperature": 0.2,
        "max_tokens": 256,
    },
    timeout=60,
)
print(resp.json()["choices"][0]["message"]["content"])

Books are the gateways to the infinite realms of human knowledge and imagination. They are not merely collections of words on pages; they are windows into the lives and minds of authors, chronicling their experiences, thoughts, and emotions. Through the pages of books, readers can embark on journeys to distant lands and times, experiencing the joys and sorrows of their characters. Books are the carriers of cultural heritage, preserving the wisdom, myths, and histories of past civilizations. They also serve as the medium for personal growth, offering insights into the human condition and the complexities of the world. Whether they are novels, biographies, science fiction, or philosophical treatises, books continue to be a vital part of our cultural and intellectual landscape, inspiring, educating, and entertaining us.


In [ ]:
!tail -n 50 /content/drive/MyDrive/mRAG_and_MSRS_source/agent_server_log/agent_runtime.log

INFO 12-02 20:18:25 api_server.py:912] vLLM API server version 0.7.3
INFO 12-02 20:18:25 api_server.py:913] args: Namespace(subparser='serve', model_tag='Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4', config='', host='127.0.0.1', port=8100, uvicorn_log_level='info', allow_credentials=False, allowed_origins=['*'], allowed_methods=['*'], allowed_headers=['*'], api_key=None, lora_modules=None, prompt_adapters=None, chat_template=None, chat_template_content_format='auto', response_role='assistant', ssl_keyfile=None, ssl_certfile=None, ssl_ca_certs=None, ssl_cert_reqs=0, root_path=None, middleware=[], return_tokens_as_token_ids=False, disable_frontend_multiprocessing=False, enable_request_id_headers=False, enable_auto_tool_choice=False, enable_reasoning=False, reasoning_parser=None, tool_call_parser=None, tool_parser_plugin='', model='Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4', task='auto', tokenizer=None, skip_tokenizer_init=False, revision=None, code_revision=None, tokenizer_revision=None, tokenizer_mode=

In [ ]:
######### Kill the Agent server

In [ ]:
#kill the agent server
!kill -TERM 8932
!fuser -k 8100/tcp

In [ ]:
!mamba run -n train_and_test python -c 'import torch; torch.cuda.empty_cache()'

In [ ]:
#resources
#11GB GPU RAM + 0.8GB additional system ram using "Qwen/Qwen2.5-3B-Instruct-GPTQ-Int4"

#Observation: need to lower more to facilitate less vram use, still need to run two models


#Trick:
'''
vllm serve Qwen/Qwen2.5-0.5B-Instruct-GPTQ-Int4 \
  --quantization gptq \
  --gpu-memory-utilization 0.5 \
  --kv-cache-dtype fp8 \
  --max-model-len 2048 \
  --tensor-parallel-size 1
'''

#Result of 50% utilization should be 5gb instead of 11, right?
#specifically using the following: --quantization gptq --gpu-memory-utilization 0.5

#8GB system,
#5.3gb gpu(which is accurate)

#Challenge, how to bring down system ram usage


In [ ]:
######### Start the Environment server

In [ ]:
# !echo "127.0.0.1" > /content/drive/MyDrive/mRAG_and_MSRS_source/environment_server_log/environment.log
# !echo "8200" >> /content/drive/MyDrive/mRAG_and_MSRS_source/environment_server_log/environment.log

# !PYTHONPATH=/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training \
#   nohup mamba run -n train_and_test vllm serve "tiiuae/Falcon3-1B-Instruct-GPTQ-Int4" --host 127.0.0.1 --port 8200 --download_dir /content/drive/MyDrive/mRAG_and_MSRS_source/vllm_cache --gpu-memory-utilization 0.3 --max-model-len 8192 --swap-space 0 --dtype=half --enforce-eager\
#   >> /content/drive/MyDrive/mRAG_and_MSRS_source/environment_server_log/environment.log 2>&1 & echo $!

In [ ]:
######### Kill the Environment server

In [ ]:
# #kill the environment server
# !kill -TERM 8710
# !fuser -k 8200/tcp

In [ ]:
# #NOTE: the cell output seems to be a lowball, but might be true, these are in kilobytes

# #NOTE: in reality seems like it takes more
# #5.2GB is the estimate from jump

# #resources stats for launched process(vram)
# !ps -p 37553 -o pid,ppid,cmd,%mem,rss,vsz
# #Uses 9.5GB on the GPU, but was supposed to use CPU right?
# #Uses a little bit of ram

In [ ]:
######### Start the Reward Server

In [ ]:
# !echo "127.0.0.1" > /content/drive/MyDrive/mRAG_and_MSRS_source/judge_server_log/judge.log
# !echo "8300" >> /content/drive/MyDrive/mRAG_and_MSRS_source/judge_server_log/judge.log

# !PYTHONPATH=/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training \
#   nohup mamba run -n train_and_test vllm serve "Qwen/Qwen2.5-0.5B-Instruct-GPTQ-Int4" --host 127.0.0.1 --port 8300 --download_dir /content/drive/MyDrive/mRAG_and_MSRS_source/vllm_cache --gpu-memory-utilization 0.15 --max-model-len 10240 --swap-space 0 --enforce-eager \
#   >> /content/drive/MyDrive/mRAG_and_MSRS_source/judge_server_log/judge.log 2>&1 & echo $!

In [ ]:
# !pkill -f 'vllm serve'

In [ ]:
# !grep -E 'MemTotal|MemFree|MemAvailable|Buffers|^Cached' /proc/meminfo

In [ ]:
######### Kill the Reward server

In [ ]:
# #kill the reward server
# !kill -TERM 9594
# !fuser -k 8300/tcp

# Using the QA pairs of yale-nlp MSRS instead of DataMorgana Generated:

In [ ]:
#NOTES:
#story_qa Benchmark from Human authored dataset SQuALITY, so is not synthetic
#meeting_qa Benchmark from Human authored dataset QMSum, so is not synthetic


story_qa = load_dataset("yale-nlp/MSRS", "story-qa")
#meeting_qa = load_dataset("yale-nlp/MSRS", "meeting-qa")


'''
#in the form
DatasetDict({
    train: Dataset({
        features: ['id', 'query', 'gold_documents', 'answer'],
        num_rows: 250
    })
    validation: Dataset({
        features: ['id', 'query', 'gold_documents', 'answer'],
        num_rows: 125
    })
    test: Dataset({
        features: ['id', 'query', 'gold_documents', 'answer'],
        num_rows: 260
    })
})
'''

#NOTICE: Merging the datasets because the train_test splitting is done elsewhere, in train_test_separate.py script
from datasets import concatenate_datasets
splits = [story_qa["train"], story_qa["validation"], story_qa["test"]]
story_qa = concatenate_datasets(splits)


#rename query column question
story_qa = story_qa.rename_columns({"query":"question"})

#For use by eventual evaluation, have the gold standard documents, they have defined gold standard documents that will encompass the answer
#story_qa_orig = story_qa

#remove columns that aren't meant to be there, only leaving questions and answers, for use as training set with mRAG
#story_qa = story_qa.remove_columns(["id","gold_documents"])


#exporting the corpus to jsonl
story_qa.to_json("/content/drive/MyDrive/mRAG_and_MSRS_source/qa_files/story_qa.json", lines=False)

train.jsonl: 0.00B [00:00, ?B/s]

dev.jsonl: 0.00B [00:00, ?B/s]

test.jsonl: 0.00B [00:00, ?B/s]

Generating train split:   0%|          | 0/250 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/125 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/260 [00:00<?, ? examples/s]

Creating json from Arrow format:   0%|          | 0/1 [00:00<?, ?ba/s]

3538237

In [ ]:


#story_qa_orig[0]
#print(story_qa_orig[0])
#print()
#print(story_qa[0])
#print(len(story_qa[0]['answer']))

#Test set, ultimately used for eval needs to actually preserve this format { id, question, gold_documents, answer}

In [ ]:
#NOTICE: number of test samples need to be changed within the file train_test_saparate.py in order to be suitable for dataset, because MSRS so small, 635 QAs
#doing 90% 10% split so update train_test_separate.py to set num_test variable = to 64

#UPDATE: added argument num_test in train_test_seperate.py so that it is easier to change on the fly

#UPDATE: outputed train.json is of different format than test.json
#train now formatted to drop 'id' and 'gold_documents' fields because subsequent script takes just queries 'question' and answer
#BUT test.json in original format because it still contains ids and ground truth/golden documents



!mamba run -n train_and_test python "/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training/utils/train_test_seperate.py" \
    --datamorgana_data_addr "/content/drive/MyDrive/mRAG_and_MSRS_source/qa_files" \
    --output_dir "/content/drive/MyDrive/mRAG_and_MSRS_source/train_test_jsons" \
    --num_test 550

# theres 650 total story_qa which we are splitting into test and train. Experiences are generated on the train split.



In [ ]:
########## Training

In [ ]:
#NOTE: Scoring using Reward model/ server

#Minimal Run, num threads = 1 and num samples = 1

#!source "/*address to the bin directory of the venv*/"
!mamba run -n train_and_test python -u "/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training/batch_training_generation_for_agent_self_training.py" \
    --queries_addr "/content/drive/MyDrive/mRAG_and_MSRS_source/train_test_jsons/train.json" \
    --output_addr "/content/drive/MyDrive/mRAG_and_MSRS_source/experiences/experience.json" \
    --num_samples 4 \
    --max_workers 16 \
    --sampling_temperature 0.7 \
    --num_shards 1 \
    --shard_id 0 \
    --no_concise \
    --save_every_n 1




SUBMIT q: Q4. Describe the setting(s) of the story, which revolves around a character with a peculiar ability who faces a life-threatening situation during a flight. Why is it important?
SUBMIT a: ['This story has two settings: first, on an airplane from San Francisco to Los Angeles and second, at the Los Angeles airport in the baggage claim and arrivals terminal. The first setting - on the airplane mid flight - is highly important to the story because it is here that the protagonist discovered the bomb in the luggage. Not only that, he discovers that bomb is on a countdown with 10 minutes remaining before detonation while the flight still has 40 minutes before arrival. It is due to this fact that the protagonist utilizes his time manipulation ability to stop the clock successfully. \n\nIn the second setting, the tensions in this story continue to rise. Despite the protagonist successfully stopping the clock in the air, it appears to continue on the ground. With both the anticipation o

In [ ]:
########### Isolate good training values

In [ ]:
# #!source "/*address to the bin directory of the venv*/"
# !mamba run -n train_and_test python "/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training/utils/extract_training_data_for_self_training.py" \
#     --input_addr "/content/drive/MyDrive/mRAG_and_MSRS_source/experiences/experience.json_0" \
#     --output_addr "/content/drive/MyDrive/mRAG_and_MSRS_source/train_test_jsons/refined_train.json"

In [ ]:
########## Train agent with good training values

In [ ]:
#because saving as 4 bit quantized after:
# !mamba run -n train_and_test pip install auto_gptq

In [ ]:
# #!source "/*address to the bin directory of the venv*/"

# #inputs_addr is single file containing inputs, a json
# !mamba run -n train_and_test python "/content/drive/MyDrive/mRAG_and_MSRS_source/inference_and_training/train_unsloth.py" \
#     --inputs_addr "/content/drive/MyDrive/mRAG_and_MSRS_source/train_test_jsons/refined_train.json" \
#     --model_addr "unsloth/Qwen2.5-0.5B-Instruct" \
#     --output_dir "/content/drive/MyDrive/mRAG_and_MSRS_source/unsloth_checkpoints" \
#     --per_device_train_batch_size 1 \
#     --gradient_accumulation_steps 64 \
#     --learning_rate 0.0001 \
#     --weight_decay 0.0 \
#     --max_steps 50 \
#     --save_steps 25 \
#     --warmup_steps 50 \
#     --max_seq_length 2048

In [ ]:
########## Testing

In [ ]:
#NOTE: this --queries_addr is different from the one passed to batch_training_generation_for_agent_self_training.py above
#this needs to be tinkered with to pass in the held out test set of the MSRS dataset since there isn't a question list provided by a competition(live rag)

#expects jsonl as input


#!source "/*address to the bin directory of the venv*/"
# !mamba run -n train_and_test python inference-and-training/batch_agent_test_day.py \
#     --queries_addr "/content/drive/MyDrive/train_test_jsons/test.json" \
#     --output_addr "/content/drive/MyDrive/test_runs/test_runs.jsonl" \
#     --max_workers 32 \
#     --num_config 1 \
#     --no_concise \
#     --agent_name "/*the address to the trained agent model*/" \





#CHANGED TO:
#python batch_agent_test_day.py \
#  --queries_addr /content/drive/MyDrive/your_tests/story_qa.json \
#  --output_addr  /content/drive/MyDrive/outputs/agent_runs.jsonl \
#  --max_workers 8 \
#  --num_config 1 \
#  --compute_metrics


#tossed in pytrec_eval to generate some eval metrics

#OR

# How to use the metric/evaluation that is done in the training script, batch_training_generation_for_agent_self_training.py, for the test script


#In Live Rag case fineweb taylored results are outputted and then graded by a judge, right?